This bit of code uses javascript to create a dynamic table of contents.

In [17]:
%%javascript
$.getScript('https://kmahelona.github.io/ipython_notebook_goodies/ipython_notebook_toc.js')

<IPython.core.display.Javascript object>

<h1 id="tocheading">Table of Contents</h1>
<div id="toc"></div>

<h1>Available GPUs</h1>

<p>See our lab infastructure page for list of machines with GPU</p>
<p>Basically we have:</p>
<ul>
<li>5 TitanX - 12 GB (shared by all across 3 terminal servers)</li>
<li>A couple other smaller 2GB ones (e.g., there's 1 on ts04)</li>

<h1>Shared GPUs usage and etiquette</h1>

<h2>Check if others are using the GPU</h2> 

<p>If they are, then you should try to use a different one</p>
<ul>
<li>You can corrupt others results if you try and access the GPU at the same time (noticed this with keras)</li>
<li>You can use too much memory and cause the job to crash if you load another model at the same time</li>
</ul>

<h3>Monitor GPU usage with the <code>nvidia-smi</code> command</h3>

<p>From the command line, run:</p>
<code>nvidia-smi</code>

<p><b>fun techincal detail!</b> <em>We can run command line arguments right from the notebook by using the exclamation point <code>!</code> as the first character</em></p>

In [7]:
!nvidia-smi

Wed Apr 26 21:54:09 2017       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 370.28                 Driver Version: 370.28                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 860M    On   | 0000:01:00.0     Off |                  N/A |
| N/A   47C    P5    N/A /  N/A |    863MiB /  2002MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU  

<p>Here we can see that we are using 863 out of 2002 MiB on the GPU</p>
<p>This indicates that the GPU is in use by someone (so we should check with whoever is using it!)</p>

<p>This also means that you should try to to be holding the GPU memory unless you are activily training! (I'm guitly of this often. If you have a Juypter Notebook open, this holds the GPU memory. So let's try to make sure we free memory when done)</p>

<h3>Use the command <code>top</code> to out who is using what</h3>
<p>Another way to find out who is using the machine...</p>
<p>From the command line run:</p>
<code>top</code>

In [12]:
# Note you'll have to stop this right away (use the stop "interrupt kernel" button at the top)
# Otherwise this will run a whole bunch.
!top

=top - 22:00:25 up  3:11,  3 users,  load average: 0.27, 0.25, 0.32
Tasks: 250 total,   1 running, 249 sleeping,   0 stopped,   0 zombie
%Cpu(s):  6.6 us,  0.6 sy,  0.4 ni, 92.3 id,  0.1 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem:  16347008 total,  6478372 used,  9868636 free,   480272 buffers
KiB Swap: 16692220 total,        0 used, 16692220 free.  3327264 cached Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND     
30259 jer       20   0 9476712 1.049g 185400 S  71.7  6.7   5:08.38 firefox     
 1223 root      20   0  435432 175696 124132 S   6.5  1.1  21:13.53 Xorg        
    1 root      20   0   33924   4364   2640 S   0.0  0.0   0:01.26 init        
    2 root      20   0       0      0      0 S   0.0  0.0   0:00.00 kthreadd    
    3 root      20   0       0      0      0 S   0.0  0.0   0:00.18 ksoftirqd/0 
    5 root       0 -20       0      0      0 S   0.0  0.0   0:00.00 kworker/0:+ 
    7 root      20   0       0      0      0 S   0.0  0.0   0:02.

<p>So you can see who (<b>jer</b>) is running what (web stuff, python, etc)!</p>
<p>You can figure out who to contact this way!</p>

<h2>Leave large GPUs free if you have small jobs</h2>
<p>Try to use a 2GB GPU if you're just doing some quick prototyping, or if your model is small. This allows others with bigger models to use the larger GPUs</p>

<h2>Eventually we will need some GPU schedule</h2>
<p>If more people are using this, it would be good to somehow let others know when, and how long</p>